### PR TITLE
fix(db): prevented invalid SQL in getWhereSQL

### LINKED ISSUE #8860: prevent easy crashes in _elgg_get_access_where_sql
In `_elgg_get_access_where_sql` there are 2 easy ways to crash your site which should be prevented by  Elgg.

First 1:
`get_access_list` could return empty array (just remove public in `access` `read` hook) causing invalid `IN` statement

Second 1:
Return empty string in `get_sql` hook causes result to be `()` which is invalid SQL.

Solution:
Just always add  the `access_id = ACCESS_PUBLIC` to the clauses after the final (get_sql) hook

Additional caveat: removing ACCESS_PUBLIC from the access_array could show all content (for example if logged out)

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #8860
- [ ] add test

### CODE REVIEW

**Review by hypeJunction — APPROVED** [review]

**Comment by jdalsem** [comment]
Do you also agree on always adding PUBLIC as the access rule if no sql is return? We could also return a empty string... but that would allow everybody to see all content... there are other ways to do that

**Comment by hypeJunction** [comment]
In theory access list should always be returning access public.

**Comment by jdalsem** [comment]
> In theory access list should always be returning access public.

yep... that is why i added the fallback

**Comment by hypeJunction** [comment]
Wouldn't it be easier to ensure that access list is never empty?

**Comment by jdalsem** [comment]
> Wouldn't it be easier to ensure that access list is never empty?

can't do... the sql hook is the last to control the sql, so it could clear all clauses... i think this fallback catches all edge cases

**Comment by hypeJunction** [comment]
Check the output of the sql hook, and append it back if it's not there. Log an error.

**Comment by jdalsem** [comment]
> Check the output of the sql hook, and append it back if it's not there. Log an error.

that would allow no one to craft their own 'valid' sql.. i don't think we want to limit that.

This PR will only make sure that the SQL is valid and has a minimal access level if no access sql is present

**Comment by hypeJunction** [comment]
Well, in that case if someone unset the public access, then there must have been a reason. Not sure if fallback the right approach. Let's see what @mrclay says

**Comment by iionly** [comment]
With walled garden enabled ACCESS_PUBLIC is unset.

**Comment by jdalsem** [comment]
True, but that still adds other acl rules, so the fallback is not used

Op 11 okt. 2016 7:50 p.m. schreef "iionly" <notifications@github.com>:

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 1]
```diff
@@ -369,6 +371,10 @@ public function getWhereSql(array $options = array()) {
 			}
 			$clauses_str .= '(' . implode(' AND ', $clauses['ands']) . ')';
 		}
+		
+		if (empty($clauses_str)) {
+			$clauses_str = "$table_alias{$options['access_column']} = " . ACCESS_PUBLIC;
```
Comment (mrclay) [inline engine/classes/Elgg/Database/AccessCollections.php]: Why not `1 = 1`?

Comment (jdalsem) [inline engine/classes/Elgg/Database/AccessCollections.php]: that is not related to access... i why do you think what i proposed is not good?

Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Because I don't really understand what we are fixing. We need tests to see the problem and the fix.

Comment (jdalsem) [inline engine/classes/Elgg/Database/AccessCollections.php]: added creating test as a todo...

Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: `1 = 0` is too extreme, it will crash the installation, because neither the site nor the user will be available.

Comment (mrclay) [inline engine/classes/Elgg/Database/AccessCollections.php]: Strawman: log an error and "undo" the effects of the **get_sql, access** hook.

**Comment by hypeJunction** [comment]
You can leave this until I get back to #10307. I will be writing exhaustive tests and mocks and will see what the expectation there should be

**Comment by jdalsem** [comment]
This has been tackled as part of https://github.com/Elgg/Elgg/pull/11155
